pyro/distributions/sine_bivariate_von_mises.py

```diff
@@ -0,0 +1,280 @@
+    assert seqs.shape == (orders, _flat_dim(vshape))
+
+    i1s = lvalues[..., :orders].T + seqs
+    assert i1s.shape == (orders, _flat_dim(vshape))
+    return i1s.view(-1, *vshape)
+
+
+class SineBivariateVonMises(TorchDistribution):
+    r""" Unimodal distribution of two dependent angles on the 2-torus (S^1 ⨂ S^1) given by
+
+    .. math::
+
+        C^{-1}\exp(\kappa_1\cos(x-\mu_1) + \kappa_2\cos(x_2 -\mu_2) + \rho\sin(x_1 - \mu_1)\sin(x_2 - \mu_2))
+
+    and
+
+    .. math::
+
+        C = (2\pi)^2 \sum_{i=0} {2i \choose i}
+        \left(\frac{\rho^2}{4\kappa_1\kappa_2}\right)^i I_i(\kappa_1)I_i(\kappa_2),
+
+    where I_i(\cdot) is the modified bessel function of first kind, mu's are the locations of the distribution,
+    kappa's are the concentration and rho gives the correlation between angles x_1 and x_2.
+
+    This distribution is helpful for modeling coupled angles such as torsion angles in peptide chains.
+    To infer parameters, use :class:`~pyro.infer.NUTS` or :class:`~pyro.infer.HMC` with priors that
+    avoid parameterizations where the distribution becomes bimodal; see note below.
+
+    .. note:: Sample efficiency drops as
+
+        .. math::
+
+            \frac{\rho}{\kappa_1\kappa_2} \rightarrow 1
+
+        because the distribution becomes increasingly bimodal.
+
+    .. note:: The correlation and weighted_correlation params are mutually exclusive.
+
+    .. note:: This distribution does not work with :class:`~pyro.infer.svi.SVI`.
+
+    ** References: **
+      1. Probabilistic model for two dependent circular variables Singh, H., Hnizdo, V., and Demchuck, E. (2002)
+
+    :param torch.Tensor phi_loc: location of first angle
+    :param torch.Tensor psi_loc: location of second angle
+    :param torch.Tensor phi_concentration: concentration of first angle
+    :param torch.Tensor psi_concentration: concentration of second angle
+    :param torch.Tensor correlation: correlation between the two angles
+    :param torch.Tensor weighted_correlation: set correlation to weigthed_corr * sqrt(phi_conc*psi_conc)
+        to avoid bimodality (see note).
+    """
+
+    arg_constraints = {'phi_loc': constraints.real, 'psi_loc': constraints.real,
+                       'phi_concentration': constraints.positive, 'psi_concentration': constraints.positive,
+                       'correlation': constraints.real}
+    support = constraints.independent(constraints.real, 1)
+    max_sample_iter = 1000
+
+    def __init__(self, phi_loc, psi_loc, phi_concentration, psi_concentration, correlation=None,
+                 weighted_correlation=None, validate_args=None):
+
+        assert (correlation is None) != (weighted_correlation is None)
+
+        if weighted_correlation is not None:
+            correlation = weighted_correlation * (phi_concentration * psi_concentration).sqrt() + 1e-8
+
+        if torch.any(phi_concentration * psi_concentration <= correlation ** 2):
```

Comment:
Can you move this check after `super().__init__(...)` and gate it by `self._validate_args`?
```py
if self._validate_args and torch.any(...):
    warnings.warn(...)
```
I think that will be safer when using the PyTorch jit.

Reply:
I'll need to read up on how jit works in PyTorch to understand why this is necessary. For now, I've moved as requested.

Reply:
Let me try to explanation of why this is necessary in more detail:
1. The PyTorch jit records tensor ops but not Python control flow. Therefore the computation in this check would be recorded, but the `if` decision would be frozen to the first iteration. Therefore the easiest solution is to simply disable this check during jitted inference.
2. The Pyro jit wrappers [disable validation globally](https://github.com/pyro-ppl/pyro/blob/dev/pyro/ops/jit.py#L92) for compiled code. Therefore we'll need to gate this line by the global validation state, which is inherited as `self._validate_args` but only after `super().__init__()` has been called.